services/director-v2/tests/unit/dynamic_scheduler/test_modules_dynamic_sidecar_scheduler__core2.py

```diff
@@ -0,0 +1,273 @@
+        app, context, workflow
+    ) as workflow_runner_manager:
+        await workflow_runner_manager.start_workflow_runner(
+            workflow_name=workflow_name, action_name="raising_error"
+        )
+        await workflow_runner_manager.wait_workflow_runner(workflow_name)
+
+        assert call_tracker.call_args_list == [
+            call(error_raising),
+            call(handle_error),
+        ]
+
+
+async def test_error_raised_but_not_handled(
+    app: AsyncMock, context: ContextInterface, workflow_name: WorkflowName
+):
+    call_tracker = AsyncMock()
+
+    class UnhandledError(RuntimeError):
+        ...
+
+    @mark_step
+    async def error_raising() -> dict[str, Any]:
+        await call_tracker(error_raising)
+        raise UnhandledError()
+
+    workflow = Workflow(
+        Action(
+            name="raising_error",
+            steps=[error_raising],
+            next_action=None,
+            on_error_action=None,
+        )
+    )
+    async with workflow_runner_manager_lifecycle(
+        app, context, workflow
+    ) as workflow_runner_manager:
+
+        await workflow_runner_manager.start_workflow_runner(
+            workflow_name=workflow_name, action_name="raising_error"
+        )
+        with pytest.raises(UnhandledError):
+            await workflow_runner_manager.wait_workflow_runner(workflow_name)
+
+        assert call_tracker.call_args_list == [
+            call(error_raising),
+        ]
+
+
+# TEST 4: test cancellation of very long pending event and schedule a new workflow which will finish
+async def test_cancellation_of_current_workflow_and_changing_to_a_different_one(
+    app: AsyncMock, context: ContextInterface
+):
+    call_tracker = AsyncMock()
+
+    # WORKFLOW_PENDING_FOREVER
+
+    @mark_step
+    async def pending_forever() -> dict[str, Any]:
+        await call_tracker(pending_forever)
+        await asyncio.sleep(1e10)
+        return {}
+
+    workflow_pending = Workflow(
+        Action(
+            name="pending",
+            steps=[pending_forever],
+            next_action=None,
+            on_error_action=None,
+        )
+    )
+
+    # WORKFLOW_FINISHES_IMMEDIATELY
+
+    @mark_step
+    async def print_something_and_finish() -> dict[str, Any]:
+        await call_tracker(print_something_and_finish)
+        print("a thing")
+        return {}
+
+    workflow_finishing = Workflow(
+        Action(
+            name="finishing",
+            steps=[print_something_and_finish],
+            next_action=None,
+            on_error_action=None,
+        )
+    )
+
+    async with workflow_runner_manager_lifecycle(
+        app=app, context=context, workflow=workflow_pending + workflow_finishing
+    ) as workflow_runner_manager:
+
+        # start the first workflow and cancel it immediately
+        await workflow_runner_manager.start_workflow_runner(
+            workflow_name="pending_workflow", action_name="pending"
+        )
+        WAIT_FOR_STEP_TO_START = 0.1
```

Comment:
you know you can also sleep for 0, which will just give the context switch that you need?